koji_wrapper/factory.py

```diff
@@ -21,7 +35,18 @@ def load_profile(cls, profile='koji', user_config=None):
     @classmethod
     def open_session(cls, profile=None, user_options=None,
                      authenticate=False, debug=False):
+        """open a koji ClientSession using profile or user_options provided
+
+        :param profile koji profile to load (optional)
+        :param user_options koji profile+user options dictionary to use
+               for opening a session
```

Comment:
Does this mean if you pass in user_options, you are required to pass in the koji profile as well? Judging by the code I would say no, so maybe this description of the param could be a bit more explicit - perhaps as simple as adding to the end something like '- this dictionary contains the same information that would be found in a profile' or something

Reply:
it's optional if you don't specify a koji profile.  So might actually be 2 different methods, maybe 3 based on the parameters passed in.

but issue here is use cases ... hmm

maybe smaller helper functions and several different outer functions

open_session(cls, profile) => returns koji.ClientSession (authenticated if possible)

_load_profile(cls, profile) => returns dictionary of options
_open_session(cls, user_options) => returns koji.ClientSession
_autenticate_session(cls, session, user_options) => attempts to authenticate user based on data passed in

and use try/except around authenticate section

Unless someone else has better idea.

Reply:
and the optional user_options could be another top level class method for different way of opening a ClientSession